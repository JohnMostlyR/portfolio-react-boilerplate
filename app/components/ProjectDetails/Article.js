import styled from 'styled-components';
import {
  BASE_FONT_BOLD,
  BASE_LINE_HEIGHT,
  MODULAR_SCALE_FACTOR,
} from '../../styles/typography';

import { theme } from '../../styles/theme';
import { SVGBackgroundImage } from '../../styles/tools';

const {
  inlineLink: { color, textShadowColor },
} = theme.projectDetails;
const H2_LINE_HEIGHT = 1.2;
const EXTERNAL_LINK_ICON = SVGBackgroundImage(`<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 576 512" aria-hidden="true">
    <path
      fill="${color}"
      d="M576 24v127.984c0 21.461-25.96 31.98-40.971 16.971l-35.707-35.709-243.523 243.523c-9.373 9.373-24.568 9.373-33.941 0l-22.627-22.627c-9.373-9.373-9.373-24.569 0-33.941L442.756 76.676l-35.703-35.705C391.982 25.9 402.656 0 424.024 0H552c13.255 0 24 10.745 24 24zM407.029 270.794l-16 16A23.999 23.999 0 0 0 384 303.765V448H64V128h264a24.003 24.003 0 0 0 16.97-7.029l16-16C376.089 89.851 365.381 64 344 64H48C21.49 64 0 85.49 0 112v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V287.764c0-21.382-25.852-32.09-40.971-16.97z"
    />
  </svg>`);

const Article = styled.article`
  max-width: 51em;

  a {
    position: relative;
    background-color: transparent;
    background-image: linear-gradient(currentColor, currentColor);
    background-origin: padding-box;
    background-position: center bottom 10%;
    background-repeat: no-repeat;
    background-size: 100% 1.5px;
    color: ${color};
    cursor: pointer;
    text-decoration: none;
    text-shadow: 3px 0 ${textShadowColor}, 2px 0 ${textShadowColor},
      1px 0 ${textShadowColor}, -1px 0 ${textShadowColor},
      -2px 0 ${textShadowColor}, -3px 0 ${textShadowColor};
    touch-action: manipulation;
    transition: color 0.1s ease-out;

    @supports (text-decoration-skip-ink: auto) {
      background: none;
      text-decoration: underline;
      text-decoration-skip-ink: auto; /* stylelint-disable-line property-no-unknown */
      text-shadow: none;
    }

    &[href^='http'] {
      padding-right: 1em;

      &::after {
        content: '';
        height: 1em;
        position: absolute;
        width: 1em;
        background-position: center;
        background-repeat: no-repeat;
        background-size: 0.75em;
        ${EXTERNAL_LINK_ICON};
      }
    }
  }

  > h2 {
    font-size: ${MODULAR_SCALE_FACTOR ** 3}rem;
    line-height: ${() => H2_LINE_HEIGHT};
    ${BASE_FONT_BOLD};
  }

  > p {
    line-height: ${BASE_LINE_HEIGHT};
    margin: ${BASE_LINE_HEIGHT}em 0;
    word-break: break-word;

    @media (min-width: 950px) {
      line-height: ${BASE_LINE_HEIGHT * 1.1};
    }
  }

  ul {
    appearance: none !important;
    margin: 0 !important;

    li {
      line-height: ${BASE_LINE_HEIGHT * 1.2};
      margin-top: 0 !important;
    }
  }
`;

export default Article;
