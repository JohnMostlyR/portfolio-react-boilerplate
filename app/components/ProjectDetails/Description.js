import styled from 'styled-components';
import { BASE_LINE_HEIGHT } from '../../styles/typography';

const Description = styled.div`
  margin: ${BASE_LINE_HEIGHT}em 0;
  max-width: 51em;
  line-height: ${BASE_LINE_HEIGHT};
  word-break: break-word;

  @media (min-width: 950px) {
    line-height: ${BASE_LINE_HEIGHT * 1.1};
  }
`;

export default Description;
