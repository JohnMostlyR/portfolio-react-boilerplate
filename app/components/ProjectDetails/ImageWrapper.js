import styled from 'styled-components';
import PropTypes from 'prop-types';
import { BASE_LINE_HEIGHT } from '../../styles/typography';

export function renderMediaBreakpoints({
  breakpoints = [],
  imageDimensions = [],
}) {
  if (breakpoints.length !== imageDimensions.length) {
    throw new Error(
      `Mismatch between the length of the breakpoints array, ${
        breakpoints.length
      } items, and the imageDimensions array, ${imageDimensions.length} items)`,
    );
  }

  const sliceOfBreakpoints = breakpoints.slice(0, -1);
  const sliceOfImageDimensions = imageDimensions.slice(1);

  return sliceOfBreakpoints.map((_breakpoint, idx) => {
    const {
      media: { size },
    } = _breakpoint;
    const { height } = sliceOfImageDimensions[idx];

    return `@media (min-width: ${size + 1}px) {
      height: ${height}px;
    };
    `;
  });
}

function getDefaultHeight(imageDimensions = []) {
  if (imageDimensions.length === 0) {
    return 'auto';
  }

  const [{ height = 'auto' }] = imageDimensions;
  return height === 'auto' ? height : `${height}px`;
}

const ImageWrapper = styled.div`
  height: ${({ imageDimensions }) => getDefaultHeight(imageDimensions)};
  margin: ${BASE_LINE_HEIGHT * 3}rem 0;
  width: 100%;
  text-align: center;

  ${props => renderMediaBreakpoints(props)};
`;

ImageWrapper.propTypes = {
  breakpoints: PropTypes.arrayOf(
    PropTypes.shape({
      media: PropTypes.shape({
        breakpoint: PropTypes.string,
        size: PropTypes.number,
      }),
      srcSet: PropTypes.string,
    }),
  ),
  imageDimensions: PropTypes.array,
};

export default ImageWrapper;
