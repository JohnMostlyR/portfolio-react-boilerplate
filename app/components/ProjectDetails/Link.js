/**
 *
 * Link
 *
 */

import styled from 'styled-components';
import { Link as RouterLink } from 'react-router-dom';
import PropTypes from 'prop-types';

const Link = styled(RouterLink)`
  display: inline-block;
  position: relative;
  color: ${({ color }) => color || 'inherit'};
  cursor: pointer;
  letter-spacing: 1px;
  padding: 0.5rem;
  text-align: left;
  text-decoration: none !important;
  white-space: nowrap;
  z-index: 2;
`;

Link.propTypes = {
  color: PropTypes.string,
};

export default Link;
