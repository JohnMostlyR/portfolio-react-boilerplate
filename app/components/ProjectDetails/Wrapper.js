import styled, { keyframes } from 'styled-components';
import PropTypes from 'prop-types';

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

const Wrapper = styled.div`
  opacity: ${({ animate }) => (animate ? 0 : 1)};
  animation-delay: 0ms;
  animation-direction: normal;
  animation-duration: 1s;
  animation-fill-mode: both;
  animation-name: ${({ animate }) => (animate ? fadeIn : 'none')};
  animation-timing-function: ease-in-out;

  @supports (display: grid) {
    display: grid;
    grid-area: mainContent;
    grid-template-areas:
      'pageHeader'
      'pageContent';
  }
`;

Wrapper.propTypes = {
  animate: PropTypes.bool,
};

Wrapper.defaultProps = {
  animate: true,
};

export default Wrapper;
