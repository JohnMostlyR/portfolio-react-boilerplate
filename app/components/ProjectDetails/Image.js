import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ImageWrapper from './ImageWrapper';

const StyledPicture = styled.picture`
  * + * {
    margin: 0;
  }
`;

function transformImagesArray(images) {
  const srcSet = {};

  images.forEach(_image => {
    const {
      file: {
        details: {
          image: { height, width },
        },
        fileName,
        url,
      },
    } = _image;
    const [relativeSizeIndication] = fileName
      .replace(/\.(png|jpe?g)/i, '')
      .trim()
      .split('-')
      .slice(-1);
    srcSet[relativeSizeIndication] = {
      dimensions: {
        height,
        width,
      },
      url,
    };
  });

  return srcSet;
}

function getImageDescription(images) {
  const [{ title = '' } = {}] = images.slice(-1);
  return title;
}

function composeSourceElements(sources) {
  return sources.map(source => {
    const {
      media: { breakpoint, size },
      srcSet,
    } = source;

    return (
      <source key={size} media={`(${breakpoint}: ${size}px)`} srcSet={srcSet} />
    );
  });
}

function Image({ images = [] }) {
  if (images.length === 0) {
    return <div />;
  }

  const { s, m, l, xl } = transformImagesArray(images);
  const altText = getImageDescription(images);
  const breakpoints = [
    {
      media: { breakpoint: 'max-width', size: 424 },
      srcSet: s.url,
    },
    {
      media: { breakpoint: 'max-width', size: 767 },
      srcSet: m.url,
    },
    {
      media: { breakpoint: 'max-width', size: 949 },
      srcSet: l.url,
    },
    {
      media: { breakpoint: 'min-width', size: 950 },
      srcSet: xl.url,
    },
  ];
  const imageDimensions = [
    s.dimensions,
    m.dimensions,
    l.dimensions,
    xl.dimensions,
  ];

  return (
    <ImageWrapper
      breakpoints={breakpoints}
      imageDimensions={imageDimensions}
      defaultImage={s.dimensions}
    >
      <StyledPicture>
        {composeSourceElements(breakpoints)}
        <img src={xl.url} alt={altText} />
      </StyledPicture>
    </ImageWrapper>
  );
}
Image.propTypes = {
  images: PropTypes.arrayOf(
    PropTypes.shape({
      file: PropTypes.shape({
        details: PropTypes.shape({
          image: PropTypes.shape({
            height: PropTypes.number.isRequired,
            width: PropTypes.number.isRequired,
          }).isRequired,
        }).isRequired,
        fileName: PropTypes.string.isRequired,
        url: PropTypes.string.isRequired,
      }).isRequired,
      title: PropTypes.string.isRequired,
    }),
  ),
};

export default Image;
