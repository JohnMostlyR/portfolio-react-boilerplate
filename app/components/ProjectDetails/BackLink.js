import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FormattedMessage } from 'react-intl';
import Link from './Link';
import messages from './messages';
import SpeechBubble from '../SpeechBubble';
import { theme } from '../../styles/theme';

const {
  borderColor: { odd: backgroundColor },
  color: { odd: textColor },
} = theme.link;
const showArrowBreakpoint = '750px';

const Wrapper = styled.div`
  @media (min-height: ${showArrowBreakpoint}) {
    margin-bottom: ${({ position }) => (position === 'top' ? 0 : '2vh')};
    margin-top: ${({ position }) => (position === 'top' ? 0 : '-7vh')};
  }
`;

/* eslint-disable jsx-a11y/anchor-is-valid */
function BackLink({ linkTo, position }) {
  return (
    <Wrapper position={position}>
      <SpeechBubble
        arrowPosition={position === 'top' ? 'bottom-right' : 'top-right'}
        arrowHeight="2vh"
        backgroundColor={backgroundColor}
        color={textColor}
        isGhost
        makeAppear
        showArrowBreakpoint={showArrowBreakpoint}
      >
        <FormattedMessage {...messages.backLink}>
          {message => (
            <Link to={linkTo} color={textColor}>
              {message}
            </Link>
          )}
        </FormattedMessage>
      </SpeechBubble>
    </Wrapper>
  );
}

BackLink.propTypes = {
  linkTo: PropTypes.string.isRequired,
  position: PropTypes.oneOf(['top', 'bottom']),
};

BackLink.defaultProps = {
  position: 'top',
};

export default BackLink;
