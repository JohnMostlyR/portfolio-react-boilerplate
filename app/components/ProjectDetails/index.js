import React from 'react';
import PropTypes from 'prop-types';
import marked from 'marked';
import Parser from 'html-react-parser';

import Wrapper from './Wrapper';
import ContentWrapper from './ContentWrapper';
import BackLink from './BackLink';
import Article from './Article';
import Image from './Image';
import Description from './Description';

import SpeechBubble from '../../components/SpeechBubble';
import { BASE_LINE_HEIGHT } from '../../styles/typography';
import { theme } from '../../styles/theme';

const { backgroundColor, color } = theme.projectDetails;

marked.options({
  breaks: true,
});

class ProjectDetails extends React.PureComponent {
  componentDidMount() {
    window.scrollTo(0, 0);
  }

  render() {
    const {
      backlink,
      project: { article, description, images, title },
    } = this.props;
    const htmlFromMD = marked(article);
    const parsedArticle = htmlFromMD.replace(
      /(<a\s+href="(https?:\/\/.+)"[^>]*>)/gi,
      (match, p1, p2) =>
        `<a href="${p2}" rel="noopener noreferrer" target="_blank">`,
    );
    const reactElementsFromHTML = Parser(parsedArticle);

    return (
      <Wrapper>
        <ContentWrapper>
          <BackLink linkTo={backlink} />
          <SpeechBubble
            arrowHeight="9vh"
            arrowPosition="bottom-right"
            backgroundColor={backgroundColor}
            color={color}
            makeAppear={false}
            padding={`${BASE_LINE_HEIGHT}rem 1rem`}
            showArrowBreakpoint="750px"
          >
            <h1>{title}</h1>
            <Description>{description}</Description>
            {images && Array.isArray(images) && <Image images={images} />}
            <Article>{reactElementsFromHTML}</Article>
          </SpeechBubble>
          <BackLink linkTo={backlink} position="bottom" />
        </ContentWrapper>
      </Wrapper>
    );
  }
}

ProjectDetails.propTypes = {
  backlink: PropTypes.string.isRequired,
  project: PropTypes.shape({
    article: PropTypes.string,
    description: PropTypes.string.isRequired,
    images: PropTypes.array,
    title: PropTypes.string.isRequired,
  }).isRequired,
};

export default ProjectDetails;
