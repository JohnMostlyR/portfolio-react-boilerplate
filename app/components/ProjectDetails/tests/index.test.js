import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import ProjectDetails from '../index';

describe('<ProjectDetails />', () => {
  const project = Object.freeze({
    article:
      '# Article for test case\n\n[add attributes to link](https://the.world/)\n',
    description: 'Description for test case',
    images: [
      {
        file: {
          details: {
            image: {
              height: 10,
              width: 20,
            },
          },
          fileName: 'test-s.png',
          url: '/test-s.png',
        },
        title: 'Test image',
      },
      {
        file: {
          details: {
            image: {
              height: 20,
              width: 30,
            },
          },
          fileName: 'test-m.png',
          url: '/test-m.png',
        },
        title: 'Test image',
      },
      {
        file: {
          details: {
            image: {
              height: 30,
              width: 40,
            },
          },
          fileName: 'test-l.png',
          url: '/test-l.png',
        },
        title: 'Test image',
      },
      {
        file: {
          details: {
            image: {
              height: 50,
              width: 60,
            },
          },
          fileName: 'test-xl.png',
          url: '/test-xl.png',
        },
        title: 'Test image',
      },
    ],
    title: 'Title for test case',
  });

  it('Should adopt the "project" property', () => {
    const wrapper = shallow(<ProjectDetails project={project} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should handle case where "images" is not provided', () => {
    const projectWithImagesNotBeingAnArray = Object.assign({}, project, {
      images: [],
    });
    const wrapper = shallow(
      <ProjectDetails project={projectWithImagesNotBeingAnArray} />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
