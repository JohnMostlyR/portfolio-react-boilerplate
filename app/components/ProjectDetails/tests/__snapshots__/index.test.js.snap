// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectDetails /> Should adopt the "project" property 1`] = `
<Wrapper
  animate={true}
>
  <ContentWrapper>
    <BackLink
      position="top"
    />
    <SpeechBubble
      arrowHeight="9vh"
      arrowPosition="bottom-right"
      backgroundColor="#aedd2c"
      color="#000"
      isGhost={false}
      makeAppear={false}
      padding="1.5rem 1rem"
      showArrowBreakpoint="750px"
    >
      <h1>
        Title for test case
      </h1>
      <Description>
        Description for test case
      </Description>
      <Image
        images={
          Array [
            Object {
              "file": Object {
                "details": Object {
                  "image": Object {
                    "height": 10,
                    "width": 20,
                  },
                },
                "fileName": "test-s.png",
                "url": "/test-s.png",
              },
              "title": "Test image",
            },
            Object {
              "file": Object {
                "details": Object {
                  "image": Object {
                    "height": 20,
                    "width": 30,
                  },
                },
                "fileName": "test-m.png",
                "url": "/test-m.png",
              },
              "title": "Test image",
            },
            Object {
              "file": Object {
                "details": Object {
                  "image": Object {
                    "height": 30,
                    "width": 40,
                  },
                },
                "fileName": "test-l.png",
                "url": "/test-l.png",
              },
              "title": "Test image",
            },
            Object {
              "file": Object {
                "details": Object {
                  "image": Object {
                    "height": 50,
                    "width": 60,
                  },
                },
                "fileName": "test-xl.png",
                "url": "/test-xl.png",
              },
              "title": "Test image",
            },
          ]
        }
      />
      <Article>
        <h1
          id="article-for-test-case"
          key="0"
        >
          Article for test case
        </h1>
        

        <p
          key="2"
        >
          <a
            href="https://the.world/"
            rel="noopener noreferrer"
            target="_blank"
          >
            add attributes to link
          </a>
        </p>
        

      </Article>
    </SpeechBubble>
    <BackLink
      position="bottom"
    />
  </ContentWrapper>
</Wrapper>
`;

exports[`<ProjectDetails /> Should handle case where "images" is not provided 1`] = `
<Wrapper
  animate={true}
>
  <ContentWrapper>
    <BackLink
      position="top"
    />
    <SpeechBubble
      arrowHeight="9vh"
      arrowPosition="bottom-right"
      backgroundColor="#aedd2c"
      color="#000"
      isGhost={false}
      makeAppear={false}
      padding="1.5rem 1rem"
      showArrowBreakpoint="750px"
    >
      <h1>
        Title for test case
      </h1>
      <Description>
        Description for test case
      </Description>
      <Image
        images={Array []}
      />
      <Article>
        <h1
          id="article-for-test-case"
          key="0"
        >
          Article for test case
        </h1>
        

        <p
          key="2"
        >
          <a
            href="https://the.world/"
            rel="noopener noreferrer"
            target="_blank"
          >
            add attributes to link
          </a>
        </p>
        

      </Article>
    </SpeechBubble>
    <BackLink
      position="bottom"
    />
  </ContentWrapper>
</Wrapper>
`;
